modules/kargoBidAdapter.js

```diff
@@ -27,58 +41,103 @@ export const spec = {
     return !!bid.params.placementId;
   },
   buildRequests: function(validBidRequests, bidderRequest) {
-    const currencyObj = config.getConfig('currency');
-    const currency = (currencyObj && currencyObj.adServerCurrency) || 'USD';
-    const bidIDs = {};
-    const bidSizes = {};
+    const currencyObj = config.getConfig(CURRENCY_KEY);
+    const currency = (currencyObj && currencyObj.adServerCurrency) ? currencyObj.adServerCurrency : null;
+    const impressions = [];
 
     _each(validBidRequests, bid => {
-      bidIDs[bid.bidId] = bid.params.placementId;
-      bidSizes[bid.bidId] = bid.sizes;
+      impressions.push(spec.getImpression(bid))
     });
 
     const firstBidRequest = validBidRequests[0];
 
-    const tdid = deepAccess(firstBidRequest, 'userId.tdid')
+    const tdid = deepAccess(firstBidRequest, TDID_BID_KEY)
 
-    const transformedParams = Object.assign({}, {
-      sessionId: spec._getSessionId(),
-      requestCount: spec._getRequestCount(),
+    const krakenParams = Object.assign({}, {
```

Comment:
Things are looking good, to bring it one step further, what do you think about extracting the functions into individual function blocks, and then tie it altogether at the end into the exported `spec` object? i.e (excerpt):
```
function isBidRequestValid(bid) {
  if (!bid || !bid.params) {
    return false;
  }

  return !!bid.params.placementId;
};

function generateUUID() {
  try {
    // crypto.getRandomValues is supported everywhere but Opera Mini for years
    var buffer = new Uint8Array(16);
    crypto.getRandomValues(buffer);
    buffer[6] = (buffer[6] & ~176) | 64;
    buffer[8] = (buffer[8] & ~64) | 128;
    var hex = Array.prototype.map.call(new Uint8Array(buffer), function(x) {
      return ('00' + x.toString(16)).slice(-2);
    }).join('');
    return hex.slice(0, 8) + '-' + hex.slice(8, 12) + '-' + hex.slice(12, 16) + '-' + hex.slice(16, 20) + '-' + hex.slice(20);
  } catch (e) {
    return '';
  }
};

function getSessionID() {
  if (!sessionId) {
    return generateUUID();
  }

  return sessionId;
};

export const spec = {
  gvlid: GVLID,
  code: BIDDER_CODE,
  isBidRequestValid,
  generateUUID,
  getSessionID,
};

registerBidder(spec);
```

Reply:
I've pulled all function declarations out of the spec object. Let me know what you think